### PR TITLE
Replace call to runtime hook  _d_arraycast with call to template object.__ArrayCast - Take 2

### PR DESCRIPTION
**Pull request by JinShil**
An attempt to finish up https://github.com/dlang/dmd/pull/8531 and set an example for a GSoC project.

~This is a WIP.  CTFE is currently running out of memory on my machine (6GB RAM) when importing `std.uni`.  I want to see how it does on the test suite.~

### CODE REVIEW

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -6715,6 +6715,42 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+            if (fromSize != toSize)
+            {
+                if ((fromSize % toSize) != 0)
+                {
+                    // A runtime check is needed in case arrays don't line up.  That check should
+                    // be done in the implementation of `object.__ArrayCast`
+
+                    // lower to `object.__ArrayCast!(TFrom, TTo)(from)`
+
+                    auto id = Identifier.idPool("__ArrayCast");
```
Comment (thewilsonator) [inline src/dmd/expressionsem.d]: should be `Identifier.__ArrayCast`, putting it in the `dmd.id.msgtable` will do that automatically.

**Comment by JinShil** [comment]
Ok, that gave me a little more information:  
```
../druntime/import/object.d(4793): Error: reinterpreting cast from `void[]` to `Array*` is not supported in CTFE
```
Anyone have any ideas?

**Comment by thewilsonator** [comment]
cc @UplinkCoder

File: src/dmd/expressionsem.d (hunk 1 above)
Comment (JinShil) [reply src/dmd/expressionsem.d]: I did put it in `msgtable`, but if I use `Identifier.__ArrayCast` I get `Error: no property __ArrayCast for type dmd.identifier.Identifier`.  What am I doing wrong?

Comment (thewilsonator) [reply src/dmd/expressionsem.d]: Derp, that should be `Id.__ArrayCast`.

File: src/dmd/expressionsem.d [hunk 2]
```diff
@@ -6715,6 +6715,42 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             return setError();
         }
 
+        if (tob.ty == Tarray && t1b.ty == Tarray && exp.e1.op != TOK.string_)
+        {
+            auto tFrom = t1b.nextOf();
+            auto tTo = tob.nextOf();
+
+            uint fromSize = cast(uint)tFrom.size();
+            uint toSize = cast(uint)tTo.size();
```
Comment (jacob-carlborg) [inline src/dmd/expressionsem.d]: `const` for these two.

**Comment by UplinkCoder** [comment]
This is deliberate. CTFE cannot assume data-layout at runtime and at ctfe will be the same.

File: src/dmd/expression.d [hunk 3]
```diff
@@ -2940,6 +2940,17 @@ extern (C++) final class ArrayLiteralExp : Expression
         return null;
     }
 
+    /**
+    Returns: the number of elements in this array literal
+    */
+    size_t length()
+    {
+        if (elements == null)
```
Comment (Basile-z) [inline src/dmd/expression.d]: `return elements == null ? 0 : elements.length;`

Comment (Basile-z) [reply src/dmd/expression.d]: Also you need to update expression.h.

**Comment by jacob-carlborg** [comment]
Tests?

**Comment by JinShil** [comment]
> Tests?

This is just replacing the runtime hook `_d_arraycast` with the template `object.__ArrayCast`.   There should be no change in functionality or behavior, though there may be changes in performance.  I'm assuming the array cast feature is already sufficiently tested, so no additional tests should be required; it should only do no harm.  `object.__ArrayCast` has unittests to verify its implementation. 

If this is merged, a followup PR will be made to druntime to remove the `_d_arraycast` runtime hook.

This does have the potential side effect of making the array cast feature available to betterC.  However, trying to tackle that simultaneously with just getting this working is taking on a little too much at a time.  I'm changing my strategy to just do a direct translation of the runtime hook, modify DMD to call it, and then follow up with additional PRs to make the feature work in betterC.  At that time, I can add the betterC tests.

**Comment by jacob-carlborg** [comment]
We had problems in the past with incorrect runtime hooks being called https://github.com/dlang/dmd/pull/9481. But I guess once `_d_arraycast` and all the tests pass it's good.

**Comment by JinShil** [comment]
OK, I'm pretty happy with this, and I've increased my understanding of how all of this works.

Please understand the scope of the PR.  We're trying to replace the runtime hooks with templates.  Please see https://forum.dlang.org/post/hequvmostuhwnvcgrcge@forum.dlang.org for justification.

I can see now that we could actually improve `object.__ArrayCast` to handle more cast, even at compile-time, but `object.__ArrayCast` was created only to replace `_d_arraycast`.  Additional PRs can be created to expand on this work once this infrastructure is in place.

Once this is merged I'll submit a few more PRs to druntime to remove `_d_arraycast` as it should no longer be needed.

**Review by thewilsonator — APPROVED** [review]

**Comment by JinShil** [comment]
Sorry for the interruption.  I just realized that since I added the `exp.e1.op != TOK.arrayLiteral` condition the check for `typeof(null)[]` is no longer required...I think.  Let's see what the test suite says.

**Comment by thewilsonator** [comment]
This introduced a regression: https://issues.dlang.org/show_bug.cgi?id=19840

**Comment by Basile-z** [comment]
This commit causes a new ICE : https://issues.dlang.org/show_bug.cgi?id=19954

**Comment by Basile-z** [comment]
This new ICE happens because during semantic `toSize` and `fromSize` are both equal to 1 so the path taken is to not write the `__ArrayCast`. But later, ruring IR, `tsize` value has changed to 16 so an `__ArrayCast` is expected.

**Comment by JinShil** [comment]
> But later, ruring IR, tsize value has changed to 16 so an __ArrayCast is expected.

Why is the size changed to 16?

**Comment by Basile-z** [comment]
I don't know. 16 is the size of an array as a fat pointer but nextof in both case is used is well used... 

Look at [the new test case](https://issues.dlang.org/show_bug.cgi?id=19954#c3). Something is seriously broken with string literals (array literals containing `char` literal don't exhibit the ICE).

**Comment by WalterBright** [comment]
@JinShil please revert this until you find the problem.

**Comment by JinShil** [comment]
@WalterBright, it has already been released and reverting it would also require reverting the druntime PR that accompanied it.  Please give me a couple of days to figure it out.

**Comment by Basile-z** [comment]
Do not start reverting, I have a working fix coming.

**Comment by Basile-z** [comment]
Are you on the fix @JinShil ? I said earlier that I would have a fix but actually I don't. There are several ways to fix the two new test cases but then either

- template9.d and foreach4.d from the test suite break
- phobos regex engine don't compile (b/c of CTFE)
- etc, depending on the way to fix.

So I give up. You'll probably handle this in a smarter way.

**Comment by JinShil** [comment]
> Are you on the fix @JinShil ?

I'm working on it and have run into the same issues in template9d and foreach4.d.  It appears we may need to special-case array literals and/or string literals (e.g. `cast(wstring)"hello"`).  I think that's probably already done somewhere in the compiler, but I didn't have time to look for it yesterday.  I'll work on it tonight.  If you or @WalterBright can think of any hints to expedite my fix, I'd be most grateful.

**Comment by Basile-z** [comment]
Valid casts are not broken. Maybe to disallow casting between arrays of different dimension would work.
People who wants to do funky stuff can still use the "cast pointer and dereference" trick or the union one.

**Comment by JinShil** [comment]
I'm still waiting on the test suite, but I think I have a fix at https://github.com/dlang/dmd/pull/10036

**Pull request merged**